File: code-coaches/medium/how-far/how_far.rb
def get_blocks_between_house_and_pond(layout)
  # Find the indices of the house and pond, and the indices of all the blocks
  house_index = layout.index("H")
  pond_index = layout.index("P")
  block_indices = layout.chars.each_index.select { |i| layout[i] == "B" }

  return block_indices.count { |block_index|
    [house_index, pond_index].min < block_index && block_index < [house_index, pond_index].max
  }
end

def main
  layout = gets.chomp
  puts get_blocks_between_house_and_pond(layout)
end

if __FILE__ == $0
  main
end

File: code-coaches/medium/how-far/HowFar.swift
func getBlocksBetweenHouseAndPond(on layout: String) -> Int {
    let houseIndex = street.firstIndex(of: "H")!
    let pondIndex = street.firstIndex(of: "P")!

    // distance(from:to:) returns the number of steps between two indices,
    // excluding the starting index (the house) but excluding the ending
    // index (the pond). We need to subtract 1 from the result to get
    // the actual number of blocks between both and get its absolute value.
    return abs(street.distance(from: houseIndex, to: pondIndex)) - 1
}

func main() {
    let layout = readLine()!
	print(getBlocksBetweenHouseAndPond(on: layout))
}

main()